<a href="https://colab.research.google.com/github/Neuralnomad-hub/Amazon-sales-data-Analysis/blob/main/codesvm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler

In [6]:
df=pd.read_csv('diabetes.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [7]:
x=df.drop('Outcome',axis=1)
y=df["Outcome"]
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=43)
model=SVC(kernel='rbf',gamma='scale',C=4)
model.fit(x_train,y_train)
predicted=model.predict(x_test)
print(predicted)

[0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 1 0 1 0 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 0 0
 1 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 0 0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 1 0 1 0 0 1 1 0 0 0 0 1 0 0 1 0 0 0 1 0
 0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 1 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 1 0 0 1 0
 0 0 0 0 0 0 0 0 0]


In [8]:
# before scaling
from sklearn.metrics import classification_report,f1_score
print(classification_report(y_test,predicted))


              precision    recall  f1-score   support

           0       0.77      0.97      0.86       152
           1       0.88      0.44      0.59        79

    accuracy                           0.79       231
   macro avg       0.82      0.71      0.72       231
weighted avg       0.81      0.79      0.77       231



In [9]:
#after scaling
scaler=StandardScaler()
x_train=scaler.fit_transform(x_train)
x_test=scaler.fit_transform(x_test)
model=SVC(kernel='rbf',gamma='scale',C=4)
model.fit(x_train,y_train)
predicted2=model.predict(x_test)
print(classification_report(y_test,predicted2))
#there fore scaling was not required as it was correctly scalled before itself

              precision    recall  f1-score   support

           0       0.77      0.89      0.82       152
           1       0.69      0.48      0.57        79

    accuracy                           0.75       231
   macro avg       0.73      0.68      0.70       231
weighted avg       0.74      0.75      0.74       231



In [10]:
# now we will use gridsearchcv for more accuracy choice of kernal,gamma and c values
params = {
    'C': [0.1, 1, 10],
    'gamma': [0.01, 0.1, 1],
    'kernel': ['rbf', 'linear']
}

grid = GridSearchCV(SVC(), params, cv=5)
grid.fit(x_train, y_train)

print(grid.best_params_)
print(grid.best_score_)


{'C': 10, 'gamma': 0.01, 'kernel': 'rbf'}
0.7782796815507096


In [11]:
model=SVC(kernel='rbf',gamma=0.01,C=10)
model.fit(x_train,y_train)
predicted1=model.predict(x_test)
print(predicted1)
print(classification_report(y_test,predicted1))


[0 0 0 0 0 0 0 1 1 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 0 1 1 0 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 0 0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 1 1 0 0 1 0 0 0 1 0 1 0 0 1 1 0 1 0 0 1 0 0 0 0 0 0 1 0
 0 0 1 1 1 0 0 0 1 0 0 0 0 1 0 0 0 0 0 1 0 0 0 1 1 0 0 1 0 0 0 0 0 0 0 0 1
 1 0 1 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 1 0 0 0 0 0 0 1 0 0 1 0
 0 0 0 1 0 0 0 0 0]
              precision    recall  f1-score   support

           0       0.78      0.92      0.85       152
           1       0.77      0.51      0.61        79

    accuracy                           0.78       231
   macro avg       0.78      0.71      0.73       231
weighted avg       0.78      0.78      0.77       231



HEnce to my best knowledge this svm algorithm was able to predict 78%..though there are various morre ways feature engineering to predict more..though we can check more possible outliers and more ways to get best possible..instead of using grid search..more best one s available out there..